### PR TITLE
Auto generate latest docs and deploy to GitHub pages

### LINKED ISSUE #258: Latest Version Documentation
Hello,

Where is the documentation of the latest version 0.4.2?

The most recent documentation I found is for 0.3 development.

Additionally, the website seems outdated as well.
The site homepage is still from version 0.2.

Could someone confirm if this library is still active or got deprecated?

Thank you!

### PR DESCRIPTION
**Pull request by kinow**
This is a WIP, that closes #258 

- [x] updated `DEPENDS-doc.txt` as required (when building with `make clean html`, it complained about `sphinx` missing, for instance, as well as `numpydoc`)
- [x] updated some auxiliary Python 2 code to use Python 3 (mainly around strings handling)
- [x] updated gitwash submodule (the commit in the submodule was using Python 2 `print`, so updated to their master commit and the build succeeded)
- [x] `apigen` doesn't appear to work with Python 2. Replaced by a fixed `api/index.txt` that uses `sphinx-autodoc` to generate the code documentation (not sure if that's OK?)
- [x] fix menu links
- [x] fix `/en/latest/` link (are we going to deploy directly to `/` and `/index.html` instead? Dropping the en/latest prefix? Or are we going to support similar structure from RTD in GH pages? cc @JDWarner)
- [x] update Travis CI to automatically publish to GitHub pages
- [ ] write docs about how to generate documentation, how publishing to GH pages works, etc.. Docs for future developers/maintainers

### CODE REVIEW

**Comment by kinow** [comment]
The build is now passing, with Python 3.8.

image

Still a few more pending items, but little by little :-)

**Comment by JDWarner** [comment]
This looks great so far!  If the issue with `apigen` is Py2, we could just drop Py2 support for the docs.  However I'm a big believer in code reuse and if Sphinx will do this now without the need to ship that ourselves, that's a very compelling argument.

The older link infrastructure existed so people could reference prior version documentation.  Scikit-Image does this, and I shamelessly borrowed that build system (the scikit-image doc build infrastructure has since evolved).  I'm open to being convinced.  I'd actually be okay deploying directly, so long as we only do so on PRs merged into master rather than all test builds.

**Comment by alex-marty** [comment]
Hi! I'm trying to use scikit-fuzzy for a project in my company. It would be very helpful to have access to updated documentation, and would certainly help spreading this project. Do you think it could be released soon? Is there anything I could do to help?

**Comment by kinow** [comment]
Hi @JDWarner 

>This looks great so far! If the issue with apigen is Py2, we could just drop Py2 support for the docs. However I'm a big believer in code reuse and if Sphinx will do this now without the need to ship that ourselves, that's a very compelling argument.

:+1: thanks. That's my understanding/opinion too. Letting Sphinx handle that, which results in dropping Py2 at least for documentation generation. Shouldn't impact Python2 support in `scikit-fuzzy`, except for this part of the build process.

>The older link infrastructure existed so people could reference prior version documentation. Scikit-Image does this, and I shamelessly borrowed that build system (the scikit-image doc build infrastructure has since evolved). I'm open to being convinced. I'd actually be okay deploying directly, so long as we only do so on PRs merged into master rather than all test builds.

The project I'm [currently working](https://github.com/cylc/cylc-flow) builds the latest docs everyday via GitHub Actions + cron. And older versions are also deployed from tags. This way users can read older versions, and also check the latest if interested in upcoming changes (that also helps developers to keep documentation up to date in their pull requests, checking that the generated docs look OK).

**Comment by kinow** [comment]
Hi @alex-marty 

>Hi! I'm trying to use scikit-fuzzy for a project in my company. It would be very helpful to have access to updated documentation, and would certainly help spreading this project. 

Same here! Though I'm using for fun, trying some examples/questions from [r/fuzzylogic](https://old.reddit.com/r/fuzzylogic/).

> Do you think it could be released soon?

I will admit I had completely forgotten this PR :grimacing: so your message served as a good reminder/nudge for me to start working on this again. Even though we live in these modern times, I usually add post-it's to my monitor, to remember what I am working on.

Just added one for this PR. So I should remember to update it today and in the coming days :crossed_fingers: 

> Is there anything I could do to help?

Definitely!!! I'm much more familiar with Javadocs & Java, and even phodocs/PHP, than Python and pydocs/sphinx.

There are pending issues in this pull request. Should you have any suggestions, or even code/fixes for these items, or for unlisted ones, feel free to comment here, or raise a PR against my branch, or we can even try to add your as contributor in my fork or in this PR (probably not doable without some github-settings-fu).

Otherwise, I will keep learning more about Sphinx as I implement the pending items whenever I have spare time and spot this new post-it when I sit in front of this old laptop  :-)

Thanks!

**Comment by kinow** [comment]
>fix /en/latest/ link (are we going to deploy directly to / and /index.html instead? Dropping the en/latest prefix? Or are we going to support similar structure from RTD in GH pages? cc @JDWarner)

#TIL about [`pathto`](https://www.sphinx-doc.org/en/master/templating.html#pathto), which links to generated pages. Just looked at the `docs/build` directory after make'ing the docs, and used `pathto('$nameoffilewithouthtmlextension')`. Seems to work fine.

I think we can build the documentation for older versions from previous tags or commits, deploy, and add a link to previous versions' docs. Solving the issue of access to the docs of previous releases.

**Comment by kinow** [comment]
> update Travis CI to automatically publish to GitHub pages

I've updated Travis CI to install the docs dependencies in the conda environment, following what was done for the test dependencies.

I couldn't find the Travis badge though, nor a project in Travis CI for `scikit-fuzzy`? I've created under my own Travis CI account, the results are here for a manually triggered build of this branch: https://travis-ci.com/github/kinow/scikit-fuzzy/builds/196978122

But even though the build passed, I think it needs further work, as the `skfuzzy` module was not found? I thought we were installing the module in the conda env, but looks like we are only building it?

image

**Comment by kinow** [comment]
Finally build passed on Travis CI. Had to instruct it to install the `skfuzzy` module, and also the docs dependencies (from `DEPENDS-docs.txt`).

image

@JDWarner should be ready for review now. Happy to update it as necessary. I haven't written any docs as that can be either part of this PR, or done later I think, in a subsequent PR.

Is there a reason we don't have `scikit-fuzzy` in Travis? Or is it there, but I just didn't find it?

Finally, the logo's link is pointing to <http://www.scikit-fuzzy.org/>. But the page is not loading for me. Is it an old domain? In which case, should we update the logo to point at the same place as the "Home" navigation link?

**Comment by JDWarner** [comment]
Thanks for circling back to this, @kinow!  I'll review as soon as possible.

Logo link should go to the same place as "Home", agreed.

Scikit-Fuzzy is on Travis, I do not know why the GitHub interface isn't working but I can see your build passed here https://travis-ci.org/github/scikit-fuzzy/scikit-fuzzy - I'll look into the integration.

**Comment by kinow** [comment]
Hi @JDWarner 

>Thanks for circling back to this, @kinow! I'll review as soon as possible.

Thanks, take your time :+1: 

>Logo link should go to the same place as "Home", agreed.

Done in the last commit!

>Scikit-Fuzzy is on Travis, I do not know why the GitHub interface isn't working but I can see your build passed here https://travis-ci.org/github/scikit-fuzzy/scikit-fuzzy - I'll look into the integration.

Oh, funny, I could swear I got a 404, but I might have misspelled something when typing the URL.

Thanks!
Bruno

**Comment by balins** [comment]
Hi @JDWarner @kinow,
I've made a [pull request](https://github.com/kinow/scikit-fuzzy/pull/1) to "docs" branch on @kinow's fork. The changes include minor fixes for docs building scripts and a Travis job that builds and publishes the latest docs from master on gh-pages branch (it will create it if it doesn't exist). We only need to generate a personal access token with "repo" access and add it as an environmental variable in Travis with key "TRAVIS_DEPLOY".

This automation makes the last checkpoint in this PR unnecessary, so we are at least one step closer to share an up-to-date documentation :smile: 

Bad news is that for some reason building the documentation doesn't work with Python 3.5 - `cd docs && make` results in
```
Exception occurred:
  File "/home/travis/miniconda/envs/test/lib/python3.5/site-packages/matplotlib/sphinxext/plot_directive.py", line 262, in setup
    app.add_directive('plot', plot_directive, True, (0, 2, False), **options)
TypeError: add_directive() got an unexpected keyword argument 'align'
```
It's hard to debug because we don't rely on specific versions of packages neither in DEPENDS.txt nor in .travis.yml when installing packages from conda.

**Comment by kinow** [comment]
Hi @balins !

Your changes look good, thanks for fixing the warning & bug, and also for trying to use Travis. I am more comfortable with GH Actions now to be honest, but I think I used Travis because it was already set up here.

@JDWarner if this is merged, it may require a few further iterations to get everything right, every Python version we want to support tested, and making sure it's deploying OK. But I think we can work on it later, with subsequent PR's if needed.

**Comment by scikit-fuzzy** [comment]
I wonder if we need to bump matplotlib.  Seems new Sphinx and old
matplotlib creates similar problems per this Issue:
https://github.com/sphinx-doc/sphinx/issues/7659

In general I'm open to merging and cleaning up in subsequent PRs.  I'll
have a look at this when I get home.

File: docs/examples/plot_tipping_problem.py [hunk 1]
```diff
@@ -177,7 +177,7 @@
 the centroid method will be used.
 
 The result is a tip of **19.84%**.
----------------------------------
+----------------------------------
```
Comment (JDWarner) [inline docs/examples/plot_tipping_problem.py:180]: I appreciate this level of detail ;)

**Review by JDWarner — APPROVED** [review]
This looks great, love the cleanup and attention to detail.  Time to get it into master and I agree with @kinow that working out minor kinks should not hold it up at this point.

**Comment by JDWarner** [comment]
Also, @kinow I think most projects are looking toward GH Actions or alternatives to Travis due to recent Travis monetization changes or throttling for free projects.  Right now we are fine on Travis, but I have concerns - if you would like to reimagine the build and deployment systems around GH Actions, perhaps in the medium term, that would have my full support.

**Comment by JDWarner** [comment]
Thanks also to @balins for prodding this (and me) back into action!

**Comment by kinow** [comment]
Thank you @JDWarner and @balins 

>Also, @kinow I think most projects are looking toward GH Actions or alternatives to Travis due to recent Travis monetization changes or throttling for free projects. Right now we are fine on Travis, but I have concerns - if you would like to reimagine the build and deployment systems around GH Actions, perhaps in the medium term, that would have my full support.

Brilliant. @balins I'll have time to work on this in the next days/week. But if you have time and would like to keep up on the momentum from your last changes, I'd be happy to help reviewing/testing, or collaborate some other way.

Bruno

**Comment by balins** [comment]
Thanks @JDWarner @kinow!
It happens that I'm working on my bachelor's thesis about fuzzy multiclass support vector machine, so you can expect more contributions from me :) It's the first open source I try to get involved in, so I will start with minor fixes, like the documentation and typos. I'll be happy to cooperate with you!
As for now, I can replace Travis with GitHub actions for this repo. I will take a closer look on dependencies and provide more repeatable environment for CI/CD.
@kinow I would rather suggest you to take care of more critical stuff involving logic of the library itself (there are still 9 open PRs!), because I don't know feel confident enough to mess up with it yet ;)

**Pull request merged**